constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -208,7 +208,80 @@ impl<T: FieldElement, V: Ord + Clone + Hash + Eq> QuadraticSymbolicExpression<T,
         }
     }
 
-    // TODO implement substitute_by_unknown where substitution is a QSE.
+    /// Substitute a variable by an unknown expression (a QuadraticSymbolicExpression).
+    pub fn substitute_by_unknown(
+        &mut self,
+        variable: &V,
+        substitution: &QuadraticSymbolicExpression<T, V>,
+    ) {
+        if !self.referenced_variables().any(|v| v == variable) {
+            return;
+        }
+
+        match &self.constant {
+            SymbolicExpression::Symbol(sym, _) if sym == variable => {
+                self.constant = SymbolicExpression::from(T::zero());
+                *self += substitution.clone();
+            }
+            _ => {
+                if let Some(mut qse) =
+                    symbolic_expression_to_quadratic_symbolic_expression(&self.constant)
+                {
+                    qse.substitute_by_unknown(variable, substitution);
+                    self.constant = SymbolicExpression::from(T::zero());
+                    *self += qse;
+                }
+            }
+        }
+
+        let mut to_add = QuadraticSymbolicExpression::from(T::zero());
+        for (var, mut coeff) in std::mem::take(&mut self.linear) {
+            if var == *variable {
+                let mut term = substitution.clone();
+                term *= &coeff;
+                to_add += term;
+            } else {
+                coeff.substitute(
+                    variable,
+                    substitution
+                        .try_to_known()
+                        .unwrap_or(&SymbolicExpression::from(T::zero())),
+                );
+                if !coeff.is_known_zero() {
+                    self.linear.insert(var, coeff);
+                }
+            }
+        }
+
+        self.quadratic.retain_mut(|(l, r)| {
+            l.substitute_by_unknown(variable, substitution);
+            r.substitute_by_unknown(variable, substitution);
+
+            match (l.try_to_known(), r.try_to_known()) {
+                (Some(lval), Some(rval)) => {
+                    to_add += (lval.clone() * rval.clone()).into();
+                    false
+                }
+                (Some(lval), None) => {
+                    to_add += r.clone() * lval.clone();
+                    false
+                }
+                (None, Some(rval)) => {
+                    to_add += l.clone() * rval.clone();
+                    false
+                }
+                _ => true,
+            }
+        });
```

Comment:
```suggestion
        std::mem::take(&mut self.quadratic)
            .into_iter()
            .for_each(|(mut l, mut r)| {
                l.substitute_by_unknown(variable, substitution);
                r.substitute_by_unknown(variable, substitution);
                to_add += l * r;
            });
```

Would this work too?

Reply:
Maybe something like this:
``` rust
self.quadratic = std::mem::take(&mut self.quadratic)
  .into_iter()
  .filter_map(|(mut l, mut r)| {
      l.substitute_by_unknown(variable, substitution);
      r.substitute_by_unknown(variable, substitution);
      match (l.try_to_known(), r.try_to_known()) {
          (Some(lval), Some(rval)) => {
              to_add += (lval * rval).into();
              None
          }
          (Some(lval), None) => {
              to_add += r * lval;
              None
          }
          (None, Some(rval)) => {
              to_add += l * rval;
              None
          }
          _ => Some((l, r)),
      }
  })
  .collect();
```
(already changed in code)